### PR TITLE
Reduce reader memory usage

### PR DESCRIPTION
**Pull request by jonashaag**
Depends on #67 

Cache trace integers, reduces reader memory usage by 30-70%

### CODE REVIEW

File: vmprof/reader.py [hunk 1]
```diff
@@ -43,12 +71,12 @@ class JittedCode(int):
...
-        return AssemblerCode(pc)
-    elif kind == VMPROF_JITTED_TAG:
-        return JittedCode(pc)
-    assert kind == VMPROF_CODE_TAG
-    return pc
+    cls = {
+        VMPROF_ASSEMBLER_TAG: AssemblerCode,
+        VMPROF_JITTED_TAG: JittedCode,
+        VMPROF_CODE_TAG: lambda x: x,
+    }[kind]
```
Comment (methane) [inline vmprof/reader.py]: Is this dict constructed for each function call?

Comment (methane) [inline vmprof/reader.py]: I feel original code is more efficient...

Comment (jonashaag) [inline vmprof/reader.py]: Please move your comments to #67

**Comment by fijal** [comment]
Does that affect PyPy or CPython memory usage?

**Comment by jonashaag** [comment]
What's a reliable way to measure PyPy memory usage? I used RSS for CPython.

**Comment by jonashaag** [comment]
OK so I did another round of benchmarking. Results: CPython ca. 20% memory reduction, 30% slowdown. PyPy: 10% reduction, 10% slowdown.

Maybe we should do some more representative benchmarks. I don't have many good example profiles (or program to profile, for that matter).

**Comment by fijal** [comment]
The integer cache is happy nonsense - let's instead make sure that we never allocate longs (so you can have negative numbers), which would make it much better. Also lists of integers are optimized on pypy

**Comment by jonashaag** [comment]
> The integer cache is happy nonsense - let's instead make sure that we never allocate longs (so you can have negative numbers), which would make it much better. Also lists of integers are optimized on pypy

I'm not sure what you're saying here. Are you talking about longs in the reader? We don't explicitly make Python allocate longs since traces are stored in `array.array` objects, which are also the best we can do in terms of compactness of representation (without doing any crazy optimizations). I take it that this also the case with PyPy.

Btw I just tried how much memory it saves if we un-dupe whole traces: In my examples roughly 10% of the traces where identical, so it's probably not worth taking that route.

**Comment by fijal** [comment]
If trace elements are all ints, then having integer cache cannot possibly make any memory savings on pypy (as pypy does not store PyInts anywhere if it's only integers in the list). The much bigger speedup would be to create a tree as we go and not create the traces one by one to later merge them into a tree (that way if we cut certain threshold we can make it run possibly in O(size-of-single-trace) memory). I'm -1 on merging such hacks if they don't (or cannot) show improvements

**Comment by jonashaag** [comment]
Fair enough. I'll close this issue then.

**Comment by fijal** [comment]
thanks. I'm really sorry for being difficult I hope I can provide more constructive criticism in the future :-) (the processing of traces bit-by-bit that I started on some random branch that I can try to find if interested is a step in a good direction I think)

**Comment by jonashaag** [comment]
No worries! This was just a cheap hack with seemingly very promising results, but you're right that it probably does more harm than good. It's not that reader memory usage is such a big problem for us.

**Pull request closed without merging**